### PR TITLE
Sniff encoding and dialect

### PR DESCRIPTION
**Pull request by mckaymatt**
There are two main things going on in the first commit. First, sniff let's a lot more data when attempting to detect the the csv dialect. The default, 1024, might not even read the header on some datasets. Second, detect the file encoding and use the correct decoder whenever we read from the file. The values are passed around so this should only need to happen once per script.

The second commit is for those wishing to run the script from another program. We don't offer a module, so we should at least allow users to send all output to stdout. cc @pprett  @Axik

### CODE REVIEW

**Comment by mckaymatt** [comment]
@Axik  Can you take a look? I probably need to add tests but at least the older tests still pass.

File: datarobot_batch_scoring/batch_scoring.py [hunk 1]
```diff
@@ -78,29 +78,61 @@ def slow_to_csv_chunk(data, header):
...
+        else:
+            self.reader = codecs.StreamRecoder(f, codecs.getencoder('utf-8'),
+                                               codecs.getdecoder('utf-8'),
+                                               codecs.getreader(encoding),
+                                               codecs.getwriter(encoding))
+
+    def __iter__(self):
+        return self
+
+    def next(self):   # python 3
```
Comment (Axik) [inline datarobot_batch_scoring/batch_scoring.py]: def __next__(self):
    return next(self.reader)

should be completely compatible for py2/3

File: datarobot_batch_scoring/batch_scoring.py [hunk 2]
```diff
@@ -120,21 +152,20 @@ def _check_for_multiline_input(self, peek_size=100):
...
     It supports multiline csv and can be a factor of 50 slower. """
 
-    def __init__(self, fd, sep=",", dialect=None, encoding='utf-8'):
-        self.fd = fd
-        self.sep = sep
-        self.dialect = dialect
-        self.reader = csv.reader(self.fd, dialect, delimiter=sep)
-        self.header = next(self.reader)
+    def __init__(self, fd, sep=",", encoding=None, dialect=None):
+        CSVReader.__init__(self, fd, sep=",", encoding=encoding,
```
Comment (Axik) [inline datarobot_batch_scoring/batch_scoring.py]: `super(self, ...)`?

File: datarobot_batch_scoring/batch_scoring.py [hunk 3]
```diff
@@ -78,29 +78,61 @@ def slow_to_csv_chunk(data, header):
...
+    def _create_reader(self):
+        fd = Recoder(self.fd, self.encoding)
+        return csv.reader(fd, self.dialect, delimiter=self.sep)
+
+
+class FastReader(CSVReader):
+    """A reader that only reads the file in text mode but not parses it. """
+
+    def __init__(self, fd, sep=",", encoding=None, dialect=None):
+        CSVReader.__init__(self, fd, sep=",", encoding=encoding,
```
Comment (Axik) [inline datarobot_batch_scoring/batch_scoring.py]: `super(self, FastReader).__init__(fd, sep=sep, encoding=encoding, dialect=dialect)`

File: datarobot_batch_scoring/batch_scoring.py [hunk 4]
```diff
@@ -226,10 +235,47 @@ def __iter__(self):
             self._ui.error('chunking {} rows took {}'.format(rows_read,
                                                              time() - t0))
 
+    def investigate_encodnig_and_dialect(self):
```
Comment (Axik) [inline datarobot_batch_scoring/batch_scoring.py]: typo here

File: datarobot_batch_scoring/batch_scoring.py [hunk 5]
```diff
@@ -78,29 +78,61 @@ def slow_to_csv_chunk(data, header):
     return buf.getvalue()
 
 
-class FastReader(object):
-    """A reader that only reads the file in text mode but not parses it. """
+class Recoder:
```
Comment (Axik) [inline datarobot_batch_scoring/batch_scoring.py]: Why it called Recorder?

File: datarobot_batch_scoring/batch_scoring.py [hunk 6]
```diff
@@ -78,29 +78,61 @@ def slow_to_csv_chunk(data, header):
...
+    Iterator that reads an encoded stream and decodes the input to UTF-8
+    for Python 2. In Python 3 the open function decodes the file.
+    """
+    def __init__(self, f, encoding):
+        f.seek(0)
+        # self.reader = codecs.getreader(encoding)(f)
+        if six.PY3:
+            self.reader = f
+        else:
+            self.reader = codecs.StreamRecoder(f, codecs.getencoder('utf-8'),
```
Comment (Axik) [inline datarobot_batch_scoring/batch_scoring.py]: is this covered by test?
Also, using codecs slowing down reader, and we want to avoid it.

**Comment by Axik** [comment]
Can you separate please stdout and encoding/dialect changes on different PR?

File: datarobot_batch_scoring/batch_scoring.py (hunk 5 above)
Comment (mckaymatt) [inline datarobot_batch_scoring/batch_scoring.py]: it re-encodes the python2 byte stream from whatever it was, to utf-8. The Python 2 csv module can only accept ascii or utf-8.

**Comment by mckaymatt** [comment]
@Axik The benchmark tests are blocked until the script can A) handle the wide datasets like the one I used for benchmarking and B) send all output to stdout. 
I have to wait on these two either way. The stdout changes are very simple.

**Comment by mckaymatt** [comment]
@pprett you mentioned that you wanted to avoiding decoding during the batching process for the sake of speed when possible.  This PR adds decoding, but I don't think it's anywhere near as significant to performance as the time added by going through the csv module. 

Here are some tests I did using a 250mb dataset, using 7 workers, against a three server cluster. The batch times vary considerably, but don't necessarily translate into longer overall time elapsed because the batching is async. 

**python 2.7 - fast mode - latest release**
/250m_bio_160a5_27_fast_xgb: [ERROR] chunking 49129 rows took 0.486748933792
/250m_bio_160a5_27_fast_xgb: [INFO] scoring complete | total time elapsed 94.6522130966s
**python 2.7 - slow mode - latest release**
/250m_bio_160a5_27_slow_xgb: [ERROR] chunking 49129 rows took 18.0388491154
/250m_bio_160a5_27_slow_xgb: [INFO] scoring complete | total time elapsed 94.5276751518s
**python 3.5 - fast mode - latest release**
/250m_bio_160a5_35_fast_xgb: [ERROR] chunking 49129 rows took 0.6467278003692627
/250m_bio_160a5_35_fast_xgb: [INFO] scoring complete | total time elapsed 94.72512340545654s
**python 3.5 - slow mode - latest release**
/250m_bio_160a5_35_slow_xgb: [ERROR] chunking 49129 rows took 14.864896297454834
/250m_bio_160a5_35_slow_xgb: [INFO] scoring complete | total time elapsed 100.5288941860199s
**python 2.7 - fast mode - this branch** 
/250m_bio_sniffbranch_27_fast_xgb: [ERROR] chunking 49129 rows took 3.38938307762
/250m_bio_sniffbranch_27_fast_xgb: [INFO] scoring complete | total time elapsed 98.6537249088s
**python 2.7 - slow mode - this branch** 
/250m_bio_sniffbranch_27_slow_xgb: [ERROR] chunking 49129 rows took 20.027875185
/250m_bio_sniffbranch_27_slow_xgb: [INFO] scoring complete | total time elapsed 93.4430410862s
**python 3.5 - fast mode - this branch** 
/250m_bio_sniffbranch_35_fast_xgb: [ERROR] chunking 49129 rows took 0.6413037776947021
/250m_bio_sniffbranch_35_fast_xgb: [INFO] scoring complete | total time elapsed 93.28744697570801s
**python 3.5 - slow mode - this branch** 
/250m_bio_sniffbranch_35_slow_xgb: [ERROR] chunking 49129 rows took 13.227560997009277
/250m_bio_sniffbranch_35_slow_xgb: [INFO] scoring complete | total time elapsed 102.19885087013245s

**Comment by Axik** [comment]
LGTM from me. @pprett ?

**Pull request merged**